### PR TITLE
[server] Serve IDE config via gitpod-protocol

### LINKED ISSUE #6601: [ide/dashboard] Assemble IDE preferences based on server-ide-configmap
Currently, the preferences page that allows the users to choose an IDE is hardcoded in the dashboard component. However, we actually have nearly everything that needs to assemble this page in the server-ide-configmap to make the available options independent from the dashboard deployments. We could add some extra values here:

https://github.com/gitpod-io/gitpod/blob/335215f565e2912d5d84db3f1502f30ed53d24fd/chart/templates/server-ide-configmap.yaml#L21-L33

e.g.:
- label/title
- logo URL
- tag like “Insiders”, “Beta”

and then loop over the available IDE entries to assemble the options. We could also add strings for the infoboxes (like the beta notice with the link to the issue/docs) to this configmap to be able to change them like this as well.

This can be implemented pretty fast and that would allow us, the IDE team, to deploy new IDEs independently of the deployment of the dashboard.

It would be very helpful to have this option pretty soon, e.g. currently we have the GoLand IDE option but the GoLand IDE image is broken since we are waiting for a new upstream version. With the proposed change, the IDE team could easily remove this option for now and add it once we have a new upstream version.

### PR DESCRIPTION
**Pull request by corneliusludmann**
## Description

With this PR the IDE entries on the preferences page are served from the `server-ide-configmap`. This allows the IDE team to add new IDEs by just editing the configmap.

## Related Issue(s)

Fixes #6601

## How to test

- Make sure that changing the IDE still works.
- You can test the code by changing the server-ide-configmap (e.g. change name, set `hidden`, remove entires etc.) and verifying that the changes are propagated to the dashboard. Just change the configmap, no component restart is needed.

## Release Notes

```release-note
NONE
```

### CODE REVIEW

**Comment by roboquat** [comment]
@corneliusludmann: Adding the "do-not-merge/release-note-label-needed" label because no release-note block was detected, please follow our [release note process](https://git.k8s.io/community/contributors/guide/release-notes.md) to remove it.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **geropl** after the PR has been reviewed.
You can assign the PR to them by writing `/assign @geropl` in a comment when ready.

Associated issue: *#6790*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/gitpod-io/gitpod/blob/main/OWNERS)**
- **[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)**
- **[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)**
- **[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: components/gitpod-protocol/src/ide-protocol.ts [hunk 1]
```diff
@@ -0,0 +1,112 @@
+/**
+ * Copyright (c) 2021 Gitpod GmbH. All rights reserved.
+ * Licensed under the GNU Affero General Public License (AGPL).
+ * See License-AGPL.txt in the project root for license information.
+ */
+
+/**
+ * `IdeServer` provides IDE related info like IDE preferences for the dashboard.
+ */
+export interface IdeServer {
```
Comment (akosyakov) [inline components/gitpod-protocol/src/ide-protocol.ts:10]: `IDEServer`

File: components/gitpod-protocol/src/ide-protocol.ts [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    /**
+     * Returns the IDE preferences.
+     */
+    getIdePreferences(): Promise<IdePreferences>;
+}
+
+/**
+ * IDE preferences (texts for the dashboard as well as available IDE options).
+ */
+export interface IdePreferences {
```
Comment (akosyakov) [inline components/gitpod-protocol/src/ide-protocol.ts:20]: `IDEPreferences`

File: components/gitpod-protocol/src/ide-protocol.ts [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+     */
+    defaultDesktopIde: string;
+
+    /**
+     * A list of available IDEs.
+     */
+    options: { [key: string]: IdeOption };
+}
+
+export interface IdeOption {
```
Comment (akosyakov) [inline components/gitpod-protocol/src/ide-protocol.ts:68]: `IDEOption`

File: components/gitpod-protocol/src/ide-protocol.ts [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+ */
+
+/**
+ * `IdeServer` provides IDE related info like IDE preferences for the dashboard.
+ */
+export interface IdeServer {
+    /**
+     * Returns the IDE preferences.
+     */
+    getIdePreferences(): Promise<IdePreferences>;
```
Comment (akosyakov) [inline components/gitpod-protocol/src/ide-protocol.ts:14]: `getIDEPreferences`

**Review by akosyakov — COMMENTED** [review]
just a minor to `IDE` instead of `ide` in type and function names

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **csweichel**, **laushinka** after the PR has been reviewed.
You can assign the PR to them by writing `/assign @csweichel @laushinka` in a comment when ready.

Associated issue: *#6790*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/gitpod-io/gitpod/blob/main/OWNERS)**
- **[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)**
- **[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)**
- **[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)**
- **[installer/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/installer/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: installer/pkg/components/server/ide-configmap.go [hunk 5]
```diff
@@ -14,32 +14,101 @@ import (
...
-			"intellij:":             common.ImageName(ctx.Config.Repository, workspace.IntelliJDesktopIDEImage, ctx.VersionManifest.Components.Workspace.DesktopIdeImages.IntelliJImage.Version),
+			"intellij":              common.ImageName(ctx.Config.Repository, workspace.IntelliJDesktopIDEImage, ctx.VersionManifest.Components.Workspace.DesktopIdeImages.IntelliJImage.Version),
 			"goland":                common.ImageName(ctx.Config.Repository, workspace.GoLandDesktopIdeImage, ctx.VersionManifest.Components.Workspace.DesktopIdeImages.GoLandImage.Version),
 		},
 		SupervisorImage: common.ImageName(ctx.Config.Repository, workspace.SupervisorImage, ctx.VersionManifest.Components.Workspace.Supervisor.Version),
+		IDEOptions: IDEOptions{
+			Options: map[string]IDEOption{
+				"theia": {
+					Title:   "Theia (legacy)",
+					Tooltip: pointer.String("This entry exists solely for legacy reasons."),
```
Comment (jankeromnes) [inline installer/pkg/components/server/ide-configmap.go:41]: This file is auto-generated, right?

**Review by jankeromnes — COMMENTED** [review]
Had a look at the dashboard and server changes, which look fine. 👍

Also briefly tested the Preferences page (there is now a brief render flash without the IDE options, as they're now loaded in a second step, but this seems like an acceptable price to pay in this case).

Thanks! 🙏

/approve

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/6811#pullrequestreview-818917786" title="Approved">jankeromnes</a>*
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **csweichel** after the PR has been reviewed.
You can assign the PR to them by writing `/assign @csweichel` in a comment when ready.

Associated issue: *#6601*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/gitpod-io/gitpod/blob/main/OWNERS)~~ [jankeromnes]
- ~~[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)~~ [jankeromnes]
- ~~[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)~~ [jankeromnes]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [jankeromnes]
- **[installer/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/installer/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by akosyakov** [comment]
/lgtm

code changes look good and it works nicely as well! great work!

**Comment by roboquat** [comment]
LGTM label has been added.  <details>Git tree hash: 8d2ab6ae8870cbf2a6be3820747986239032b62a</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/6811#issuecomment-982619013" title="LGTM">akosyakov</a>*, *<a href="https://github.com/gitpod-io/gitpod/pull/6811#pullrequestreview-818917786" title="Approved">jankeromnes</a>*
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign **csweichel** after the PR has been reviewed.
You can assign the PR to them by writing `/assign @csweichel` in a comment when ready.

Associated issue: *#6601*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/gitpod-io/gitpod/blob/main/OWNERS)~~ [akosyakov,jankeromnes]
- ~~[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)~~ [jankeromnes]
- ~~[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)~~ [jankeromnes]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [jankeromnes]
- **[installer/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/installer/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: installer/pkg/components/server/ide-configmap.go (hunk 5 above)
Comment (corneliusludmann) [reply installer/pkg/components/server/ide-configmap.go:41]: No, that's a duplicate of the content of the configmap helm template. As far as I understood it, this part of the installer will replace the configmap template and will be the single point of truth eventually.

**Comment by corneliusludmann** [comment]
Ah, a word about backwards compatibility: The `server-ide-configmap` needs to be deployed first and is backwards compatible with the old server. However, the server is not backwards compatible with the old configmap. I will take care that we deploy the new configmap right after we merged this to main so that the meta team can deploy the server whenever they are ready for it.

I have deliberately refrained from making the server compatible with the old configmap since this would make the server source code more cumbersome. And deploying the new configmap can be done pretty fast with nearly no risk of rollback.

**Comment by corneliusludmann** [comment]
/assign @MrSimonEmms 

Would be great if you could approve the installer part as we already discussed it during the installer office hours.

**Comment by akosyakov** [comment]
>  I will take care that we deploy the new configmap right after we merged this to main so that the meta team can deploy the server whenever they are ready for it.

Does it mean that the current server will just fail and reject them? I have not tried it.

**Comment by MrSimonEmms** [comment]
/lgtm

@corneliusludmann that looks good. Hopefully this will just approve the installer work

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/6811#issuecomment-982619013" title="LGTM">akosyakov</a>*, *<a href="https://github.com/gitpod-io/gitpod/pull/6811#pullrequestreview-818917786" title="Approved">jankeromnes</a>*, *<a href="https://github.com/gitpod-io/gitpod/pull/6811#issuecomment-982675751" title="LGTM">MrSimonEmms</a>*

Associated issue: *#6601*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/gitpod-io/gitpod/blob/main/OWNERS)~~ [MrSimonEmms,akosyakov,jankeromnes]
- ~~[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)~~ [jankeromnes]
- ~~[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)~~ [jankeromnes]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [jankeromnes]
- ~~[installer/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/installer/OWNERS)~~ [MrSimonEmms]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Pull request merged**